### ISSUE TITLE
OL v7

### ISSUE THREAD

**Issue by paul121** [farmOS/farmOS-map#178]
Tracking for bumping the OL dependency to v7: https://github.com/openlayers/openlayers/releases/tag/v7.0.0

I noticed a couple breaking changes that may effect us:

> The forEachLayerAtPixel method has been removed from the Map class.

I know we use this downstream in farmOS, likely in farmOS-map too?

> Deprecated options have been removed from the Overlay component.

Might not directly effect us but should check on the ol-popup library which uses an overlay.

**Comment by paul121** [comment]
I only see one usage of the `map.forEachLayerAtPixel` method in farmOS-map core: https://github.com/farmOS/farmOS-map/blob/e90a9f8f3f36cb027a5ccdcec65b9a95c22b935a/examples/simple-html-consumer/static/index.html#L42

We use it much more in farmOS core for the "popup" behavior.

**Comment by mstenta** [comment]
@paul121 We don't use `forEachLayerAtPixel()` in either farmOS core or farmOS-map core, we use `forEachFeatureAtPixel()`, which hasn't been deprecated.

`forEachFeatureAtPixel()` is still part of OL 7.1.0: https://openlayers.org/en/latest/apidoc/module-ol_Map-Map.html#forEachFeatureAtPixel

**Comment by paul121** [comment]
The [upgrade notes](https://github.com/openlayers/openlayers/blob/v10.4.0/changelog/upgrade-notes.md) include all breaking changes and deprecations for the following major and minor versions:

- [v7](https://github.com/openlayers/openlayers/releases/tag/v7.0.0)
- [v8](https://github.com/openlayers/openlayers/releases/tag/v8.0.0) 
- [v9](https://github.com/openlayers/openlayers/releases/tag/v9.0.0)
- [v10](https://github.com/openlayers/openlayers/releases/tag/v10.0.0)

**Comment by paul121** [comment]
Things I am testing (verified they work unless noted otherwise):

- [x] Current location control
- [x] Geocoder control
- [x] Snapshot
- [x] Rotation/rotate to North
- [x] Layer switcher (multiple layer types... testing all that farmOS core provides)
- [x] Sidepanel w/ layer switcher
- [x] Edit controls
- [x] Resizing map
- [ ] Snapping grid - **the grid is displayed, but cursor does not snap to the grid in OL 10.4**

**Comment by paul121** [comment]
> the Snapping grid does not have any "snapping" functionality... The grid is created, but when using the editor to draw a geometry, it does not snap to the grid. Not sure which change might have caused this, still need to look closer

Looking closer it seems like this might be an issue with the Edit control. I discovered that the Edit control was snapping to geometries that are drawn in the current "session", but if you stopped drawing (press escape) and then resumed drawing again after clicking a drawing tool a second time, you could no longer snap to the previously drawn geometries in the map. A bit of an odd issue, but I also confirmed that this is not an issue in the current version of farmOS-map.

I was able to fix this by making sure we always re-load vector features from all layers when starting/resuming drawing - simply un-indenting this logic! Unfortunately this didn't enable snapping of the Snapping grid point features. https://github.com/farmOS/farmOS-map/blob/280e316c0df89653cc00d21f9eedb3a104765edf/src/control/Edit/Edit.js#L390-L404

So... the other difference I'm learning is that the Snapping Grid's `ol-grid` interaction is creating an "unmanaged" vector layer when it calls [`layer.setMap(map/null)`](https://github.com/symbioquine/ol-grid/blob/8e5c4ae8bc69427c1cadc8d64b42c05dc05a611d/src/index.js#L166)

See docs on [`layer.map`](https://openlayers.org/en/latest/apidoc/module-ol_layer_Vector-VectorLayer.html) property: 
> Sets the layer as overlay on a map. The map will not manage this layer in its layers collection, and the layer will be rendered on top. This is useful for temporary layers. The standard way to add a layer to a map and have it managed by the map is to use [map.addLayer()](https://openlayers.org/en/latest/apidoc/module-ol_Map-Map.html#addLayer).

Basically, this layer won't be picked up by the [`map.getLayerGroup()`](https://openlayers.org/en/latest/apidoc/module-ol_Map-Map.html#getLayerGroup) method that we are calling in the Edit control's `enableSnap` logic. So... perhaps `ol-grid` could be updated to create a layer that is managed by the map? 🤔 (curious your thoughts @symbioquine )

I'm still not sure what exactly changed in OL to affect these things... but getting closer to understanding why :-)

**Comment by paul121** [comment]
> the Snapping grid does not have any "snapping" functionality... The grid is created, but when using the editor to draw a geometry, it does not snap to the grid. Not sure which change might have caused this, still need to look closer

Oh, also worth noting that the snapping grid is able to snap to vector features when drawing the initial two points to orient the grid. It uses logic very similar to the Edit control to do this.

**Comment by symbioquine** [comment]
> Basically, this layer won't be picked up by the [`map.getLayerGroup()`](https://openlayers.org/en/latest/apidoc/module-ol_Map-Map.html#getLayerGroup) method that we are calling in the Edit control's `enableSnap` logic. So... perhaps `ol-grid` could be updated to create a layer that is managed by the map, instead of this unmanaged layer overlay thing? 🤔 (curious your thoughts [@symbioquine](https://github.com/symbioquine) )

That's by design. The snapping grid control in farmOS-map [adds its own snap interaction](https://github.com/farmOS/farmOS-map/blob/280e316c0df89653cc00d21f9eedb3a104765edf/src/control/SnappingGrid/SnappingGrid.js#L156-L161).

Notice in the edit control code how there isn't really any connection between the draw/modify interactions and the snap interaction (they're both just independently added to the map). In OL v6 (and my understanding) that meant that the snapping grid's snap interaction just needed to be added to the map - and have the correct control point features added (to the snap interaction that is).

My guess is that something either changed with how multiple snap interactions coexist or something broke with the logic that the snapping grid uses to maintain its own snap interaction... I'll try to take a look. Any chance you can push your changes so far so I can pick up with the snapping grid part where you've gotten to?

**Comment by paul121** [comment]
> That's by design. The snapping grid control in farmOS-map [adds its own snap interaction](https://github.com/farmOS/farmOS-map/blob/280e316c0df89653cc00d21f9eedb3a104765edf/src/control/SnappingGrid/SnappingGrid.js#L156-L161).

> My guess is that something either changed with how multiple snap interactions coexist or something broke with the logic that the snapping grid uses to maintain its own snap interaction... 

Ah that makes sense. It's interesting that the snapping behavior within the Edit control somewhat broke as well.

Opened a PR here #206
